src/libs/wcautil/WixToolset.WcaUtil/qtexec.cpp

```diff
@@ -112,20 +132,73 @@ static HRESULT HandleOutput(
     BOOL bFirst = TRUE;
     BOOL bUnicode = TRUE;
     HRESULT hr = S_OK;
+    OVERLAPPED overlapped;
+    HANDLE rghHandles[2];
+    BOOL fRes = TRUE;
+    DWORD dwRes = ERROR_SUCCESS;
+
+    ::ZeroMemory(&overlapped, sizeof(overlapped));
 
     // Get buffer for output
     pBuffer = static_cast<BYTE *>(MemAlloc(OUTPUT_BUFFER, FALSE));
     ExitOnNull(pBuffer, hr, E_OUTOFMEMORY, "Failed to allocate buffer for output.");
 
-    while (0 != dwBytes)
+    overlapped.hEvent = ::CreateEvent(nullptr, TRUE, FALSE, nullptr);
+    ExitOnNullWithLastError((overlapped.hEvent && (overlapped.hEvent != INVALID_HANDLE_VALUE)), hr, "Failed to create event");
+    
+    rghHandles[0] = hProcess;
+    rghHandles[1] = overlapped.hEvent;
+    
+    fRes = ::ConnectNamedPipe(hRead, &overlapped);
+    if (!fRes)
+    {
+        dwRes = ::GetLastError();
+        if (dwRes == ERROR_IO_PENDING)
+        {
+            dwRes = ::WaitForSingleObject(overlapped.hEvent, INFINITE);
+            ExitOnNull((dwRes == WAIT_OBJECT_0), hr, HRESULT_FROM_WIN32(dwRes), "Failed to wait for process to connect to stdout");
+            fRes = TRUE;
+        }
+        else if (dwRes == ERROR_PIPE_CONNECTED)
+        {
+            fRes = TRUE;
+        }
+        ExitOnNullWithLastError(fRes, hr, "Failed to connect to stdout");
+    }    
+
+    while (true)
     {
         ::ZeroMemory(pBuffer, OUTPUT_BUFFER);
-        if (!::ReadFile(hRead, pBuffer, OUTPUT_BUFFER - 1, &dwBytes, NULL) && GetLastError() != ERROR_BROKEN_PIPE)
+        fRes = ::ResetEvent(overlapped.hEvent);
+        ExitOnNullWithLastError(fRes, hr, "Failed to reset event");
+
+        fRes = ::ReadFile(hRead, pBuffer, OUTPUT_BUFFER - 1, nullptr, &overlapped);
+        if (!fRes)
+        {
+            dwRes = ::GetLastError();
+            if (dwRes == ERROR_BROKEN_PIPE)
+            {
+                break;
+            }
+            ExitOnNullWithLastError((dwRes == ERROR_IO_PENDING), hr, "Failed to wait for stdout data");
+        }
+
+        dwRes = ::WaitForMultipleObjects(ARRAYSIZE(rghHandles), rghHandles, FALSE, INFINITE);
+        // Process terminated, or pipe abandoned
+        if ((dwRes == WAIT_OBJECT_0) || (dwRes == WAIT_ABANDONED_0) || (dwRes == (WAIT_ABANDONED_0 + 1)))
+        {
+            break;
```

Comment:
If the user wants the output of the process, then this is not correct. It is possible that the process exits before all the output is read.

You'll probably need to make it wait only a specific amount of time when reading from the pipe. Then it would stop reading when either the process exited and it timed out or the pipe was broken. You should be able to use `AppWaitForSingleObject` with `ExitOnWaitObjectFailure` to help with knowing when it timed out.

Reply:
Once the process has terminated, attempting to read from the pipe would yield ERROR_BROKEN_PIPE.
However, the process is bound to terminate after it has finished writing to the pipe, so I don't think that's really an issue

Reply:
> attempting to read from the pipe would yield ERROR_BROKEN_PIPE

If that was true, then there would have been no issue. If the process does not close the handles then it will not yield ERROR_BROKEN_PIPE. Unless you're saying that the changes you made is now making the process close them?

> However, the process is bound to terminate after it has finished writing to the pipe

Yes, that's true. But that doesn't mean we have read all of the data from the pipe. If the user cares about the output, then we need to make sure we have read everything.

Reply:
See https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-exitprocess
Windows implicitly closes all the object handles of the process in step 6, before signaling the process object (step 8).
Thus by the time WaitForMultipleObjects returns with the process handle, the pipe had been closed already and would yield ERROR_BROKEN_PIPE if read from. This is also in line with tests I've been running

Reply:
What's the point of reading asynchronously then? Trying to read synchronously from a broken pipe shouldn't hang.

Reply:
When the process returns with inheriting its pipes to other processes, then you're right.
However when the process inherits its pipes to other processes and exits, the wait on the pipe alone hangs. Waiting for either the process or the pipe fixes that

Reply:
A synchronous read from a broken pipe will never hang. So if you're encountering an issue with the current code where it's hanging, then that means the pipe is not broken. I don't understand what you're trying to tell me. Either your changes have magically made it where the pipe now gets broken in which case we can keep the synchronous read. Or the pipe is still not being broken in which case we need to properly code the asynchronous reading such that the full output is captured. This code will not capture the full output because it gives up reading as soon as the process exits. There is no guarantee that we have read all the data from the pipe when the process exits because the OS can buffer the data in the pipe such that the code that is writing to the pipe can compete before the data is read.

Reply:
The problem arises when the process creates child processes of its own and they inherit the same stdout/stderr handles.
In that case even after the first process exits, the pipe handles are kept alive until the child processes exit as well. In practice instead the installation waits for the entire process tree.
The fix handles that case by waiting for either the process to terminate, or the pipe.

Reply:
This function is doing two things. The first is waiting for the process to exit. Your updated code does this correctly. The second is capturing all of the output from that process. Your updated code does not do this.

Reply:
What makes you think that it doesn't capture the output? In my tests the output is logged the same as it was before the change. Did you see a test where the log was truncated?
Both the existing code and the PR check for ERROR_BROKEN_PIPE as a sign that the pipe has been closed on the write side, so it seems that both ways are at the same level of risk to miss out on some data. That said, I haven't seen logs where the process log was truncated

Reply:
This is a classic race condition. It will work under some conditions and fail under other conditions. You should be able to force it to fail by sleeping at the beginning of the loop (line 166) and minimizing the size of the read buffer (`OUTPUT_BUFFER`).

Reply:
I agree with that. However, it is the same with the existing implementation
The PR doesn't increase the risk of loosing data, yet it fixes an indefinite hang in the described scenario

Reply:
There is no race condition in the existing implementation. You have introduced it by breaking out of the loop when the process exited.

Reply:
I disagree with that- the same race condition exists implicitly
Once the process exits, any unread data in the pipe is discarded. 
You can see in the existing code that error ERROR_BROKEN_PIPE is silently ignored after ReadFile. You have no guarantee that there weren't some unlogged remains there.

Reply:
No. The OS will continue to return the data that it buffered from a broken pipe. It will only return ERROR_BROKEN_PIPE when trying to read from an empty broken pipe.

Reply:
What makes you think that? Did you see it documented?

Reply:
You can see the concept of buffering documented in https://learn.microsoft.com/en-us/windows/win32/ipc/anonymous-pipe-operations. It would be non-sensical for the buffer to be discarded so unless you can find documentation for that you should assume that it's actually possible to read all of the data from a pipe without having to synchronize the reader and writer to not close their end of the pipe before the other is done.

Reply:
I added a PeekNamedPipe call to read any remains in the pipe after the process has ended